{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contents\n",
    "\n",
    "* [How to use mtSet from command line](mtSet_scripts.ipynb)\n",
    "    * [Preprocessing](mtSet_preprocess.ipynb)\n",
    "    * [Phenotype Simulator](mtSet_phenosim.ipynb)\n",
    "    * [Running mtSet](mtSet_analyze.ipynb)\n",
    "    * [Postprocessing](mtSet_postprocess.ipynb)\n",
    "    * [Example for command line usage](example_usage.ipynb)\n",
    "* [How to use mtSet within python](mtSet_python.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#How to use mtSet within python\n",
    "\n",
    "This tutorial shows how to use mtSet within python. For a tutorial on how to use mtSet from the command line using the limix scripts (mtSet_preprocess, mtSet_analyze, mtSet_postprocess, mtSet_simPheno) please refer to the tutorials:\n",
    "* [Tutorial on how to use mtSet from command line](mtSet_scripts.ipynb)\n",
    "* [Example for command line usage](example_usage.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Setting up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "notes"
    }
   },
   "outputs": [],
   "source": [
    "# activiate inline plotting\n",
    "%matplotlib inline\n",
    "\n",
    "from setup import get_1000G_mtSet\n",
    "import scipy as sp\n",
    "import scipy.linalg\n",
    "import limix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ./../data/1000g/chrom22_subsample20_maf0.10.bed exsits\n",
      "File ./../data/1000g/chrom22_subsample20_maf0.10.bim exsits\n",
      "File ./../data/1000g/chrom22_subsample20_maf0.10.fam exsits\n",
      "File ./../data/1000g/pheno.phe exsits\n",
      "File ./../data/1000g/chrom22.cov exsits\n"
     ]
    }
   ],
   "source": [
    "# loading 1000G genotypes for mtSet demo\n",
    "get_1000G_mtSet()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# base name for bed, bim and fam\n",
    "bfile = './../data/1000g/chrom22_subsample20_maf0.10'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split genotypes into regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from limix.mtSet.core import plink_reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import genotype positions\n",
    "bim = plink_reader.readBIM(bfile,usecols=(0,1,2,3))\n",
    "chrom = bim[:, 0].astype(float)\n",
    "pos = bim[:, -1].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# uses splitter to split the genotypes\n",
    "from limix.mtSet.core.splitter import Splitter\n",
    "split = Splitter(pos=pos,chrom=chrom)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The method _splitGeno_ allows to define the regions that will then considered for the analysis with mtSet.\n",
    "Information relative to the calculated regions can be cached in an external file by activating the cache option (see below).\n",
    "\n",
    "| Argument        | Default       | Datatype | Explanation |\n",
    "| ------------- |:-------------:|:--------:| --------|\n",
    "| _method_      | 'slidingWindow' | str | Uses a sliding window approach to define regions (a region-based approach will be availabe soon) |\n",
    "| _size_      | 5E+04 (50kb) | float | Window size. Pace is set at half the size of the window |\n",
    "| _minSnps_      | 1 | int | Windows with number of SNPs lower that this threshold are not considered |\n",
    "| _maxSnps_      | sp.inf | int | Windows with number of SNPs higher that this threshold are not considered |\n",
    "| _cache_       | False | bool | If true, it activates the caching |\n",
    "| _out_dir_     | './cache' | str | outdir of the cache file |\n",
    "| _fname_       | None | str | Name of the file |\n",
    "| _rewrite_     | False | bool | If true and the cache file already exists, the cache file is overwritten |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1380 windows\n"
     ]
    }
   ],
   "source": [
    "split.splitGeno(cache=True, fname='regions.h5', minSnps=4)\n",
    "print '%d windows' % split.nWindows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply mtSet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this paragraph we showcase how to construct the mtSet class that will then be used for the set test analysis.\n",
    "We showcase both the full mtSet that models relatedness as random effect by the means of an individual-to-individual covariance matrix and the approximated model (mtSetPC) that models relatedness as fixed effect using principal component from the covariance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# import phenotype and sample relatedness\n",
    "pheno_file = './../data/1000g/pheno.phe'\n",
    "sample_relatedness_file = './../data/1000g/chrom22.cov'\n",
    "Y = sp.loadtxt(pheno_file)\n",
    "R = sp.loadtxt(sample_relatedness_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# compute eigenvalues and eigenvectors of the sample relatedness matrix\n",
    "S_R, U_R = scipy.linalg.eigh(R) # these are needed for the full mtSet model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# caculate fixed effects with rel\n",
    "F = U_R[:, ::-1][:, :10] #Â it considered the first 10 PCs\n",
    "F = sp.concatenate([F, sp.ones((F.shape[0], 1))], 1) # add an intercept term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if 0:\n",
    "    # use full mtSet implementation\n",
    "    # (relatedness is modeled as random effect by the means of\n",
    "    # an individual-to-individual covariance matrix)\n",
    "    mtSet = limix.MTSet(Y, S_R=S_R, U_R=U_R)\n",
    "else:\n",
    "    # use mtSetPC\n",
    "    # (relatedness s modelled as fixed effect\n",
    "    # using principal component from the covariance)\n",
    "    mtSet = limix.MTSet(Y, F=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Null model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the analysis is parallelized across different set of regions and permutations,\n",
    "it might be convenient to cache the results from the optimization of the null model\n",
    "(as the null model need to be optimized only once).\n",
    "\n",
    "| Argument        | Default       | Datatype | Explanation |\n",
    "| ------------- |:-------------:|:--------:| --------|\n",
    "| _cache_       | False | bool | If true, it activates the caching |\n",
    "| _out_dir_     | './cache' | str | outdir of the cache file |\n",
    "| _fname_       | None | str | Name of the file |\n",
    "| _rewrite_     | False | bool | If true and the cache file already exists, the cache file is overwritten |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "RV = mtSet.fitNull()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The returned dictionary contains:\n",
    "* B: value of the optimized effect sizes\n",
    "* Cg: value of the genetic trait-to-trait covariance\n",
    "* Cn: value of the residual trait-to-trait covariance\n",
    "* conv: bool that indicates convergence of the optimization\n",
    "* time: time elpased for optimizing the parameters\n",
    "* NLL0: negative log likelihood of the null model\n",
    "* LMLgrad: norm of the gradient of the negative log likelihood dividived by the number of parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# read fam\n",
    "bim = plink_reader.readBIM(bfile,usecols=(0,1,2,3))\n",
    "fam = plink_reader.readFAM(bfile,usecols=(0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".. window 0 - (22, 16025000-16075000) - 21 snps\n",
      ".. window 1 - (22, 16050000-16100000) - 23 snps\n",
      ".. window 2 - (22, 16125000-16175000) - 7 snps\n",
      ".. window 3 - (22, 16225000-16275000) - 9 snps\n",
      ".. window 4 - (22, 16250000-16300000) - 16 snps\n",
      ".. window 5 - (22, 16275000-16325000) - 12 snps\n",
      ".. window 6 - (22, 16325000-16375000) - 5 snps\n",
      ".. window 7 - (22, 16350000-16400000) - 5 snps\n",
      ".. window 8 - (22, 16475000-16525000) - 8 snps\n",
      ".. window 9 - (22, 16500000-16550000) - 7 snps\n"
     ]
    }
   ],
   "source": [
    "n_wnds = 100 #Â only hundred windows are considered\n",
    "LLR = sp.zeros(n_wnds) # vector with test statistics of the n_wnds regions\n",
    "for wnd_i in range(n_wnds):\n",
    "    wnd_pos = split.wnd_pos[wnd_i]\n",
    "    nSnps = split.nSnps[wnd_i]\n",
    "    idx_wnd_start = split.idx_wnd_start[wnd_i]\n",
    "    print '.. window %d - (%d, %d-%d) - %d snps' % (wnd_i, wnd_pos[0], wnd_pos[1], wnd_pos[2], nSnps)\n",
    "    \n",
    "    Xr = plink_reader.readBED(bfile, useMAFencoding=True, start = idx_wnd_start, nSNPs = nSnps, bim=bim , fam=fam)['snps']\n",
    "\n",
    "    # multi trait set test fit\n",
    "    RV = mtSet.optimize(Xr)\n",
    "    LLR[wnd_i] = RV['LLR'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The returned dictionary from _mtSet.optimize_ contains:\n",
    "* Cr: value of the region-term trait-to-trait covariance\n",
    "* Cg: value of the genetic trait-to-trait covariance\n",
    "* Cn: value of the residual trait-to-trait covariance\n",
    "* variances: n_traits-by-3 matrix of variance explained by the three contributions (region, background, noise) for the  traits\n",
    "* conv: bool that indicates convergence of the optimization\n",
    "* time: time elpased for optimizing the parameters\n",
    "* NLLAlt: negative log likelihood of the alternative model\n",
    "* LLR: test statistics\n",
    "* LMLgrad: norm of the gradient of the negative log likelihood dividived by the number of parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### P-values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "P values are obtained from a relatively small number of genome-wide permutations, fitting a parametric model to the null distribution. Here we showcase the permutation procedure by considering 10 permutations for the 10 regions analyzed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "permutation 0\n",
      "permutation 1\n",
      "permutation 2\n",
      "permutation 3\n",
      "permutation 4\n",
      "permutation 5\n",
      "permutation 6\n",
      "permutation 7\n",
      "permutation 8\n",
      "permutation 9\n"
     ]
    }
   ],
   "source": [
    "n_perms = 10\n",
    "LLR_null = [] # in this list test statistics from permutations will be stored\n",
    "for perm_i in range(n_perms):\n",
    "    \n",
    "    #1. generate permutation\n",
    "    print 'permutation %d' % perm_i\n",
    "    sp.random.seed(perm_i)\n",
    "    perm_idxs = sp.random.permutation(Y.shape[0])\n",
    "    \n",
    "    #2. scan on the 100 regions\n",
    "    for wnd_i in range(n_wnds):\n",
    "        wnd_pos = split.wnd_pos[wnd_i]\n",
    "        nSnps = split.nSnps[wnd_i]\n",
    "        idx_wnd_start = split.idx_wnd_start[wnd_i]\n",
    "        Xr = plink_reader.readBED(bfile, useMAFencoding=True, start = idx_wnd_start, nSNPs = nSnps, bim=bim , fam=fam)['snps']\n",
    "        Xr = Xr[perm_idxs, :] # permute samples in region term\n",
    "        RV = mtSet.optimize(Xr)\n",
    "        LLR_null.append(RV['LLR'][0])\n",
    "LLR_null = sp.array(LLR_null)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The parametric fit to the ditribution of the test statistics under the null and the consequent conversion of the observed test statistics in P values is performed by the module _limix.stats.chi2mixture_ as shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from limix.stats.chi2mixture import Chi2mixture\n",
    "c2m = Chi2mixture(tol=4e-3)\n",
    "c2m.estimate_chi2mixture(LLR_null)\n",
    "pv = c2m.sf(LLR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x10fe3f490>]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAENCAYAAAABh67pAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEbtJREFUeJzt3W+MHPV9x/H35nbvzq4Nh88VdW3Hp9pWgTZQaGOMCPIh\nUGPfAyzSPmiJiApRy5PYqapU4KqS/ahVpaoJdlSKQkGpWmGpoFoQwaV2yglEgRRhG0Nj5HNj1QbF\nBZtr7dbGd8v2wcxx6/P+vZ2d3fnt+yWNbv7d7Feju8/O/OY3MyBJkiRJkiRJkiRJkiRJkqQmPQmc\nBo7UWGcUOAi8A0y0vyRJUpLuAG6metAPAe8Cq+Lp5WkUJUlqzOcaWOcV4OMay+8DngVOxdMftVqU\nJCk5jQR9PeuBZcBLwJvA/QlsU5KUkHwC2ygAtwB3AYuB14DXgWMJbFuS1KIkgv4kUXPNhXh4GbiJ\nK4N+ElibwOdJUi85DqxL44NGqH4x9jrgANBHdER/BLihwnqltlSWTbs6XUAX2dXpArrIrk4X0EV2\ndbqALtJydjZyRP80sImoN81JYCdRcw3A48BRYBx4G/gU+B7w760WJklKRiNB/7sNrPOX8SBJ6jJJ\n9LpR8yY6XUAXmeh0AV1kotMFdJGJThfQafn88NjAwJrxTtfRLNvoJakB+fzw2KJFv3psxYo/LJFA\ndnpEL0ldpq9vyfahobsT62lj0EtSl8nl+geT3J5BL0ldplS6dDHJ7Rn0ktRlisXzu6emDkx2uo6F\n8GKsJDUo6nUz8iIZy85MFStJXcJeN5Kk2gx6SQqcQS9JgTPoJSlwBr0kBc6gl6TAGfSSFDiDXpIC\nZ9BLUuAMekkKnEEvSYFrJOifBE4DR+qs90VgBvhKq0VJkpLTSNA/BWyus04f8BfAOJBrtShJUnIa\nCfpXgI/rrLMNeAb4sOWKJEmJSqKNfiWwFXgsnvZxxJLURZII+u8AjxAFfA6bbiSpq+QT2MavA3vj\n8eXAFmAaeK7CurvKxifiQZI0ZzQeUjdC/V43EF24rdbrxiYdqUtFr61bMz44uH5iYGDNeD4/PNbp\nmvSZlrOzkSP6p4FNREfrJ4GdQCFe9nirBUjqrHx+eKxQWPHo0NDd62bnTU0dWAswM3Pmhc5VpqSk\n2Z4+24YvqYsMDKwZX7bs3i/Pn3/27L7xTz45saUTNekyLWend8ZKPS6X6x+sPL+wKO1a1B4GvdTj\nSqVLFyvPn76Qdi1qD4Ne6nHF4vndU1MHJsvnTU3tP14sntvTqZqULNvoJZHPD4/19S3dlssVFpVK\n0xeKxXN7vBDbNTKVnXavlKTmtZydNt1IUuAMekkKnEEvSYEz6CUpcAa9JAXOoJekwBn0khQ4g16S\nAmfQS1LgDHpJCpxBL0mBM+glKXAGvSQFzqCXpMA1EvRPAqeBI1WWfxU4DLwNvArcmExpkqS03AHc\nTPWgvw24Oh7fDLxeZT2fRy9JzUstO0eoHvTlrgFOVVlm0EtS87ruxSNfB3z9mCR1kXyC27oTeBC4\nvcY6u8rGJ+JBkjRnNB5SN0LtppsbgUlgXY11bLqRpOZ1RRv954lCfmOdbRj0ktS8lrMz18A6TwOb\ngOVE3Sx3AoV42ePAE8C9wH/G86aBDRW2U2rw8yRJczKVnR7RS1Lzuq7XjSSpyxj0khQ4g16SAmfQ\nS1LgDHpJCpxBL0mBM+glKXAGvSQFzqCXpMAZ9JIUOINekgJn0EtS4Ax6SQqcQS9JgTPoJSlwSb4z\nVkpVPj881te3ZHsu1z9YKl26WCye3z0zc8aX00vzGPTKpHx+eKxQWPHo0NDdn72neGrqwFoAw166\nnE03yqS+viXby0MeYGjo7nV9fUu3daqmTsrnh8cGBtaMDw6unxgYWDOezw+PdbomdQ+P6JVJuVz/\nYOX5hUVp19Jpnt2onkaO6J8kein4kRrr7AaOAYeBmxOoS6qpVLp0sfL86Qtp19Jpnt2onkaC/ilg\nc43lY8A6YD3wB8BjCdQl1VQsnt89NXVgsnze1NT+48XiuT2dqqlTPLtRPY003bwCjNRYfg/w/Xj8\nDWAIuJboLEBqi9kmibNn923L5QqLSqXpC8XiuT292FTh2Y3qSaKNfiVwsmz6FLCKBIPebnSqZGbm\nzAv+HXx2drP28jb63jy7UWVJXYzNzZsuVVlvV9n4RDzU5IUmqTbPboIzGg+JmR/Q1YwAzwNfqLDs\nb4gCe288fRTYxJVH9KUmPu8zAwNrxpctu/fL8+efPbtv/JNPTmxpdnuSlDELys5ySfSjfw74Wjy+\nEZgiwWYbLzRJUmsaabp5mugIfTlRW/xOoBAvexx4gajnzSTwv8ADSRbohSZJak1LpwNNWtDpR+U2\n+v3Hp6d/tt02SEk9oOWmm64PepjtdbPUC02SelHLQZ+maj1xJEnVtZydPtRMkgLnQ83Us7wRT73C\noFdP8kY89RKbbtSTfOKjeklPHNF7iq75vBFPvST4oPcUXZV4I556SfBNN56iqxKfZ69eEvwRvafo\nqsQnPqqXBB/0nqKrGp9nr14RfNONp+iS0pbPD48NDKwZHxxcPzEwsGY8nx8e62g9nfzwNHiKLilN\n3dgBJBMPNZOkrGjDy5K64sUjkqRYN3YAMeglKUHd2AHEoJekBHVjBxDb6CUpYQm/LClT2emLRySp\neam8eGQzcBQ4BjxcYfnVwPPAIeAd4PdaLUqSlJ4+YBIYAQpEYX79vHX+BPjzeHw5cIbK/fM9opek\n5rX9iH4DUdCfAKaBvcDWeet8ClwVj19FFPQzrRYmSUpGvaBfCZwsmz4Vzyv3XeAG4APgMPDNxKqT\nJLWs3iMQGjll2Ay8BdwJrAX2AzcB5yqsu6tsfCIeJElzRuMhNRuB8bLpHVx5QfYHwO1l0z8CfqPC\ntmyjl6TmtT0788Bxooux/VS+GPvXwM54/Fqi5p1lFbZl0EtS81LJzi3Ae0QXZXfE8x6KB4AVwA+B\nt4EjwH1VtmPQS1LzMpWdmSpWkrpEKjdMSZIyzKCXpMAZ9JIUOINekgJn0EtS4Ax6SQqcQS9JgTPo\nJSlwBr0kBc6gl6TAGfSSFDiDXpICZ9BLUuAMekkKnEEvSYEz6CUpcAa9JAXOoJekwBn0khS4RoJ+\nM3AUOAY8XGWdUeAg8A4wkURhkqR09AGTwAhQAA4B189bZwh4F1gVTy+vsi1fDi5JzWv7y8E3EAX9\nCWAa2AtsnbfOfcCzwKl4+qNWi5IkJade0K8ETpZNn4rnlVsPLANeAt4E7k+sOklSy/J1ljdyylAA\nbgHuAhYDrwGvE7Xpz7erbHwC2/Mlab7ReEhMvaB/H1hdNr2auSaaWSeJmmsuxMPLwE3UD3pJ0pUm\nuPwgeGe7PzAPHCe6GNtP5Yux1wEHiC7cLgaOADdU2JYXYyWpealk5xbgPaKLsjvieQ/Fw6xvEfW8\nOQJsr7Idg16Smpep7MxUsZLUJdrevVKSlHEGvSQFzqCXpMAZ9JIUOINekgJn0EtS4Ax6SQqcQS9J\ngTPoJSlwBr0kBc6gl6TAGfSSFDiDXpICZ9BLUuAMekkKnEEvSYEz6CUpcAa9JAXOoJekwDUS9JuB\no8Ax4OEa630RmAG+kkBdkqSU9AGTwAhQAA4B11dZ71+AHwC/VWVbvhxckprX9peDbyAK+hPANLAX\n2FphvW3AM8CHrRYkSUpWvaBfCZwsmz4Vz5u/zlbgsXjaI3dJ6iL5OssbCe3vAI/E6+bioZpdZeMT\n8SBJmjMaD6nZCIyXTe/gyguy/wH8NB7OAaeBeypsyyN9SWpe27MzDxwnuhjbT/WLsbOeonqvG4Ne\nkprXcnbWa7qZAb4B/JCoZ83fAj8BHoqXP95qAZKkcHhEL0nNa3v3SklSxhn0khQ4g16SAmfQS1Lg\nDHpJCpxBL0mBM+glKXAGvSQFzqCXpMAZ9JIUOINekgJn0EtS4Ax6SQqcQS9JgTPoJSlwBr0kBc6g\nl6TAGfSSFLhGg34zcBQ4BjxcYflXgcPA28CrwI2JVCdJSkUfMAmMAAXgEHD9vHVuA66OxzcDr1fY\nju+MlaTmpZKdtwHjZdOPxEM11wCnKsw36CWpeam8HHwlcLJs+lQ8r5qvAy+0UpQkKTn5BtZp5tvk\nTuBB4PaFlSNJSlojQf8+sLpsejWVm2ZuBL5H1Eb/cZVt7Sobn4gHSdKc0XhIVR44TnQxtp/KF2M/\nT3TBdmON7dhGL0nNSy07twDvEYX5jnjeQ/EA8ARwBjgYDz+usA2DXpKa13J25pKookGllD8vSPn8\n8Fhf35LtuVz/YKl06WKxeH73zMwZL35L4Wo5Oxtpo1eXyOeHxwqFFY8ODd29bnbe1NSBtQCGfWf4\nxassMOgzpK9vyfbykAcYGrp73dmz+7YZLunzi1dZ4bNuMiSX6x+sPL+wKO1aVP2Lt69v6bZO1SRV\nYtBnSKl06WLl+dMX0q5FfvEqOwz6DCkWz++emjowWT5vamr/8WLx3J5O1dTL/OJVVthGnyGz7b5n\nz+7blssVFpVK0xeKxXN7bA/ujPiLd+3lbfR+8ar72L1SakHU62apX7xqp0xlpzdMSVLzUnl6pSQp\nwwx6SQqcQS9JgTPoJSlwBr0kBc6gl6TAGfSSFDiDXpICZ9BLUuAMekkKnEEvSYFrJOg3A0eBY8DD\nVdbZHS8/DNycTGmSpDT0AZPACFAADgHXz1tnDJh9Wt+twOtVtuVDzeaMdrqALjLa6QK6yGinC+gi\no50uoIu0/aFmG4iC/gQwDewFts5b5x7g+/H4G8AQcG2rhQVutNMFdJHRThfQRUY7XUAXGe10ASGp\nF/QrgZNl06fiefXWWdV6aZKkJNQL+kZPGeY/FN9mGknqEvVeJfg+sLpsejXREXutdVbF8+Y7jl8A\n5XZ2uoAu4r6Y476Y476IHG/3B+TjDxkB+ql/MXYj1S/GSpK61BbgPaKLsjvieQ/Fw6zvxssPA7ek\nWp0kSZKk1jwJnAaO1FhnFDgIvANMlM1v5GasLGllXzTyu1my0H2xGngJeDeev71tFaZnoftikKjL\n8qF4/q52FZiSVv4/ILqv5yDwfBtq64RW9scJ4O142Y/bUt08dxDdDVut2CGif9rZLpfL45+N3IyV\nNQvdF438btYsdF/8AvBr8fgSombDXv67WBz/zBNd/7q1HQWmpJX9APBHwD8Az7WluvS1sj9+Cixr\n9IOSeNbNK8DHNZbfBzzLXG+dj+KfjdyMlTUL3ReN/G7WLHRf/IzoSx/gPPAT4BfbUWCKWvm7+L/4\nZz/RAdGniVeXnlb2wyqijh9PcGV37qxqZX9AE/shjYearSf65nkJeBO4P57fyM1Yoam2L3pRI/ti\nhOiI5430yuqIWvvic0RffKeBfwb+LfXq0lNrP3wb+GOy/UXXrFr7o0T09/Am8Pv1NlSvH30SCkQ9\nce4iOg19jegUtBf71FfbF8c6WVSH1NsXS4BngG8SHdmHrNa++JSoKetq4J+AXyE6nQ9Rtf3wy8B/\nEbVHj3aquA6o9XfxJeAD4OeB/UTXOl+ptqE0gv4k0SnHhXh4GbiJ6Ai+3s1Yoam2L3ox6GvtiwLR\nKevfA/s6VWCKGvm7+G+iI7vNhBv01fbDLUTP1BojukB9FfB3wNc6U2Zqav1dfBCv8yHRAcAGagR9\nUkaofkHhOuAA0cXXxfF6N9DYzVhZNELz+6KR382iEZrfFzmif+Jvp1BfmkZofl8sJ7ogB7CI6B99\nrK1Vtt8IC///ANhEOL1uYGH7YzGwNF7n54BXgd9sa5XA00TfLpeIvoEe5Mobqr5FdBRyhMu7y1W6\nGSvLWtkXs7/7Sfy7D6RQbzstdF98iai54hDRqfpBoqPYLFvovvgC8BbRjYhHgD9Nqd52aeX/Y9Ym\nwul1s9D98UtE/x+z3W5DyE5JkiRJkiRJkiRJkiRpTjMPK/wr5nqavUdYj0SRpGAt9GGF3yB6ro8k\nKQNGuDzo1wIvEj2n5mWiRzzM969Ejz6QJGXACJcH/Y+AdfH4rfF0uTVEN1Il9pTONJ51I0mKLAFu\nA/6xbF7/vHV+J17eiw9+lKRMGmHuiP4q5h5IVs1bwMYkC0jjefSSpMj/EL0d6rfj6RxwY9ny64Br\niB5HLEnKgPkPLnuA6Aj/RaKHkr3L5Q+r2wn8WbolSpIkSZIkSZIkSZIkSZIkSZIkSVKG/D/3lghL\nYMvl1QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x110369990>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#makes a manhattan plot\n",
    "wnd_chrom = split.wnd_pos[:n_wnds,0]\n",
    "wnd_start = split.wnd_pos[:n_wnds,1]\n",
    "wnd_end = split.wnd_pos[:n_wnds,2]\n",
    "import pylab as pl\n",
    "pl.plot(wnd_start, -sp.log10(pv), 'o', color='MidnightBlue')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
